src/main/haskell/matching-logic/src/Kore/MatchingLogic/Signature/Poly.hs

```diff
@@ -0,0 +1,172 @@
+with that signature.
+-}
+module Kore.MatchingLogic.Signature.Poly
+  (SignatureInfo(..),PolySignature) where
+  -- ,ValidatedSignature,fromValidated,findLabel) where
+import           Data.Char                 (isAlpha, isAlphaNum)
+import           Data.Coerce
+import           Data.Functor.Foldable
+import           Data.Map.Strict           (Map)
+import qualified Data.Map.Strict           as Map
+import           Data.Proxy
+import           Data.Set                  (Set)
+import qualified Data.Set                  as Set
+import           Data.Text                 (Text)
+import qualified Data.Text                 as Text
+
+import           Data.Reflection
+import           Data.Text.Prettyprint.Doc
+
+import           Kore.MatchingLogic.AST
+
+data SortPat = PApp Text [SortPat] | Var Int
+  deriving Show
+
+data SignatureInfo =
+  MkSignatureInfo
+  { sortCons :: Map Text Int
+  , labels   :: Map Text (Int,SortPat,[SortPat])
+  }
+  deriving Show
+
+newtype ValidatedSignature =
+  MkValidatedSignature SignatureInfo
+instance Show ValidatedSignature where
+  show (MkValidatedSignature sig) = show sig
+
+fromValidated :: ValidatedSignature -> SignatureInfo
+fromValidated (MkValidatedSignature info) = info
+
+data PolySignature s
+
+sortTermValid :: Map Text Int -> Int -> SortPat -> Bool
+sortTermValid sortCons nargs t = check t
+  where check (Var v) = 0 <= v && v < nargs
+        check (PApp c ts) =
+          case Map.lookup c sortCons of
+            Just arity -> arity == length ts && all check ts
+            Nothing    -> False
+
+isValid :: SignatureInfo -> Bool
+isValid sigInfo =
+  all (\(arity,r,args) ->sortOk arity r && all (sortOk arity) args) (labels sigInfo)
+ where sortOk arity t = sortTermValid (sortCons sigInfo) arity t
+
+validate :: SignatureInfo -> Maybe ValidatedSignature
+validate sig = if isValid sig then Just (MkValidatedSignature sig) else Nothing
+
+data RawSort = RawSort Text [RawSort]
+  deriving Eq
+data RawLabel = RawLabel Text [RawSort]
+  deriving Eq
+
+prettyName :: Text -> Doc ann
+prettyName name
+  | isAlpha (Text.head name) && Text.all isAlphaNum name = pretty name
+  | otherwise = pretty (show name)
+
+prettyPolyTerm :: (Pretty t) => Text -> [t] -> Doc ann
+prettyPolyTerm name args = prettyName name <> braced (map pretty args)
+
+instance Show RawSort where
+  showsPrec _ (RawSort con args) s = shows con (showsBraces args s)
+instance Pretty RawSort where
+  pretty (RawSort con args) = prettyName con
+
+showsBraces :: (Show a) => [a] -> ShowS
+showsBraces items s = '{':showsItems items ('}':s)
+   where showsItems [] s          = s
+         showsItems [x] s         = shows x s
+         showsItems (x:l@(_:_)) s = shows x (',':showsItems l s)
+
+instantiate :: [RawSort] -> SortPat -> RawSort
+instantiate args (Var v)          = args !! v
+instantiate args (PApp con cargs) = RawSort con (map (instantiate args) cargs)
+
+findSort1 :: forall s . (Reifies s ValidatedSignature) =>
```

Comment:
I'm not sure if I made this comment already or not, but the names for the (`findSort1`, `findSort`) pair could be better. How about (`sortFromName`, `resolveSort`) or something similar?

The same for `findLabel`.